vllm/v1/sample/logits_processor.py

```diff
@@ -0,0 +1,491 @@
+      or self.peek_removed_if_can() are read in a step, no new removals
+      are registered using self.removed_append()
+    * Elements of self._removed are never directly modified, added or
+      removed (i.e. modification is only via self.removed_append() and
+      self.pop_removed_if_can())
+    
+    Guarantees under above assumptions:
+    * self.removed is always sorted in descending order
+    * self.pop_removed_if_can() and self.peek_removed_if_can() both return
+      the lowest removed request index in the current step
+    """
+
+    _removed: list[RemovedRequest]
+    _is_removed_sorted: bool
+    moved: list[MovedRequest]
+    added: list[AddedRequest]
+
+    def __init__(
+        self,
+        removed: Optional[list[RemovedRequest]] = None,
+        moved: Optional[list[MovedRequest]] = None,
+        added: Optional[list[AddedRequest]] = None,
+    ) -> None:
+        self._removed = removed or []
+        self.moved = moved or []
+        self.added = added or []
+        self._is_removed_sorted = False
+
+    def _sort_removed(self) -> None:
+        """Sort removed request indices in
+        descending order.
+        
+        Idempotent after first call in a
+        given step, until reset.
+        """
+        if not self._is_removed_sorted:
+            self._removed.sort(reverse=True)
+            self._is_removed_sorted = True
+
+    @property
+    def removed(self) -> list[RemovedRequest]:
+        """Removed request indices sorted in
+        descending order"""
+        self._sort_removed()
+        return self._removed
+
+    def removed_append(self, index: int) -> None:
+        """Register the removal of a request from
+        the persistent batch.
+
+        Must not be called after the first time
+        self.removed, self.pop_removed_if_can() or
+        self.peek_removed_if_can() are invoked.
+        
+        Args:
+          index: request index
+        """
+        if self._is_removed_sorted:
+            raise RuntimeError("Cannot register new removed request after"
+                               " self.removed has been read.")
+        self._removed.append(index)
+
+    def has_removed(self) -> bool:
+        return bool(self._removed)
+
+    def peek_removed_if_can(self) -> Optional[int]:
+        """Return lowest removed request index"""
+        if self.has_removed():
+            self._sort_removed()
+            return self._removed[-1]
+        return None
+
+    def pop_removed_if_can(self) -> Optional[int]:
+        """Pop lowest removed request index"""
+        if self.has_removed():
+            self._sort_removed()
+            return self._removed.pop()
+        return None
+
+    def reset(self):
+        """Reset batch info at end of step"""
+        self._removed = []
+        self._is_removed_sorted = False
+        self.moved = []
+        self.added = []
+
+    def buildBatchUpdate(self, batch_size: int) -> BatchUpdate:
```

Comment:
Can combine these into single method

```suggestion
    def get_and_reset(self, batch_size: int) -> BatchUpdate:
```

Reply:
Done